mlflow/sklearn/__init__.py

```diff
@@ -455,6 +461,265 @@ def load_model(model_uri):
+        }
+        self.eval_dataset_info_map = defaultdict(lambda: defaultdict(list))
+        self.metric_api_call_arg_dict_list_map = defaultdict(list)
+
+    def call_scope(self, scope_key):
+        class CallScope:
+            def __enter__(inner_self):
+                self.scope_flag_dict[scope_key] = True
+
+            def __exit__(inner_self, exc_type, exc_val, exc_tb):
+                self.scope_flag_dict[scope_key] = False
+
+        return CallScope()
+
+    def get_run_id_for_model(self, model):
+        return self.model_id_to_run_id_map.get(model._mlflow_uuid, None)
+
+    def should_log_eval_metrics(self, metric_value=None):
+        """
+        Test whether we should log metric. will check metric value types (if provided)
+        and scope flags
+        """
+        return (np.isscalar(metric_value) if metric_value is not None else True) and \
+            not self.scope_flag_dict[self.fit_scope_key] and \
+            not self.scope_flag_dict[self.eval_and_log_metrics_scope_key] and \
+            not self.scope_flag_dict[self.model_score_scope_key]
+
+    def register_model(self, model, run_id):
+        """
+        In `patched_fit`, we need register the model with the run_id used in `patched_fit`
+        So that in following metric autologging, the metric will be logged into the registered
+        run_id
+        """
+        model._mlflow_uuid = uuid4()
+        self.model_id_to_run_id_map[model._mlflow_uuid] = run_id
+
+    @staticmethod
+    def gen_name_with_index(name, index):
+        assert index >= 0
+        if index == 0:
+            return name
+        else:
+            return f'name_{index + 1}'
+
+    @staticmethod
+    def gen_dataset_sample_rows(eval_dataset):
+        n_sample = max(len(eval_dataset), INPUT_EXAMPLE_SAMPLE_ROWS)
+        return eval_dataset[:n_sample]
+
+    def register_eval_dataset(self, model, eval_dataset):
+        """
+        Register eval dataset into eval_dataset_info_map, it will do:
+         1. inspect eval dataset var name.
+         2. check whether eval_dataset_info_map already registered this eval dataset.
+            will check by object id and comparing row samples.
+            Note: only check object id is not sufficient, id may be reused.
+         3. register eval dataset (with id and row samples information)
+         4. return eval dataset name with index.
+        """
+        eval_dataset_name = _inspect_original_var_name(eval_dataset)
+        eval_dataset_id = id(eval_dataset)
+        if eval_dataset_name is None:
+            eval_dataset_name = 'unknown_dataset'
+
+        registered_dataset_list = self.eval_dataset_info_map[model._mlflow_uuid][eval_dataset_name]
+
+        row_samples = self.gen_dataset_sample_rows(eval_dataset)
+        index = len(registered_dataset_list)
+        for i, id_i, row_samples_i in enumerate(registered_dataset_list):
+            if eval_dataset_id == id_i and np.array_equal(row_samples, row_samples_i):
```

Comment:
What happens if the datasets are pandas dataframes, which are supported by sklearn? does `enumerate(<pandas_dataframe>)` return a sequence of numpy arrays?

Reply:
Looks like it doesn't:

```python
>>> import pandas as pd
>>> df = pd.DataFrame({"a": [1, 2, 3], "b": [4, 5, 6]})
>>> [x for x in df]
['a', 'b']
```

Reply:
seems here code logic doesn't related to pandas.
the `registered_dataset_list` is always a list (contains elements of tuple `(id, dataset: Union[np.ndarray, pandas.DataFrame])`
the `enumerate` will simply enumerate the list.

Reply:
@WeichenXu123 got it, but what happens when we run `np.array_equal()` on pandas dataframes on line 596?

Reply:
I remove the `np.array_equal(row_samples, row_samples_i)` checking.
I think we compare variable name and variable ID should be sufficient.
The case variable names are the same and variable (eval dataset) ID being reused, such thing is almost impossible to happen.